src/app/devices/kvm/kvm.component.ts

```diff
@@ -13,38 +13,103 @@ import { AfterViewInit, Component, ElementRef, HostListener, Input, OnInit, View
 export class KvmComponent implements OnInit, AfterViewInit {
   @ViewChild('canvas', { static: false }) canvas: ElementRef | undefined
   public context!: CanvasRenderingContext2D
-  // logger: ConsoleLogger | null = null
-
-  // constructor () { }
 
   // setting a width and height for the canvas
   @Input() public width = 400
   @Input() public height = 400
+  @Input() deviceUuid: string = ''
+  @Input() showKvm: boolean = false
+  @Output() showKvmChange = new EventEmitter<boolean>()
   module: any
+  redirector: any
+  dataProcessor!: IDataProcessor | null
+  mouseHelper!: MouseHelper
+  keyboardHelper!: KeyBoardHelper
+  logger!: ILogger
+  kvmState: number = 0
+  powerState: any = 0
+  btnText: string = 'Disconnect'
+  isPoweredOn: boolean = false
 
-  ngOnInit (): void {
-    // this.logger = new ConsoleLogger(1)
+  constructor (private readonly devicesService: DevicesService) {
 
   }
 
+  ngOnInit (): void {
+    this.logger = new ConsoleLogger(1)
+  }
+
   ngAfterViewInit (): void {
+    this.instantiate()
+    // this.setAmtFeatures()
+    this.autoConnect()
+  }
+
+  checkPowerStatus (): boolean {
+    return this.powerState.powerstate === 2
+  }
+
+  instantiate (): void {
     this.context = this.canvas?.nativeElement.getContext('2d')
+    this.redirector = new AMTKvmDataRedirector(this.logger, Protocol.KVM, new FileReader(), this.deviceUuid, 16994, '', '', 0, 0, '13.76.223.84:3000/relay')
+    this.module = new AMTDesktop(this.logger as any, this.context)
+    this.dataProcessor = new DataProcessor(this.logger, this.redirector, this.module)
+    this.mouseHelper = new MouseHelper(this.module, this.redirector, 200)
+    this.keyboardHelper = new KeyBoardHelper(this.module, this.redirector)
 
-    // this.module = new AMTDesktop(this.logger as any, this.context)
+    this.redirector.onProcessData = this.module.processData.bind(this.module)
+    this.redirector.onStart = this.module.start.bind(this.module)
+    this.redirector.onNewState = this.module.onStateChange.bind(this.module)
+    this.redirector.onSendKvmData = this.module.onSendKvmData.bind(this.module)
+    this.redirector.onStateChanged = this.onConnectionStateChange
+    this.module.onProcessData = this.dataProcessor.processData.bind(this.dataProcessor)
+    this.module.onSend = this.redirector.send.bind(this.redirector)
   }
 
-  @HostListener('mouseup')
-  onMouseup (): void {
-    console.log('mouseup')
+  autoConnect (): void {
+    if (this.redirector !== null) {
```

Comment:
should be "!=" to ensure not undefined as well

Reply:
Updated